File: 0525-contiguous-array/0525-contiguous-array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
       
        for i in range(len(nums)) :
            if nums[i]==0:
                 nums[i] = -1
        

        print(nums)



        # step 2 use map and check the largest window with sum 0 

        my_dict = {}
        
        # my_dict[0] = 0

        sum =0
        ans =0
        i =0

        for i in range(len(nums)):
            sum += nums[i]
            print(f"i = {i}, sum = {sum}")
            if sum ==0:
                ans = max(ans, i +1)
            elif sum in my_dict:
                ans = max(ans,i-my_dict[sum])
            else:
                my_dict[sum]=i
        


        return ans



        # step 3 return ans 

File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        // important and good question redo it once more by yourself
        unordered_map<int,int> mymap;
        
        int sum = 0;
        int longest_subarray = 0;
        
        for(int i=0;i<nums.size();++i)
        {
            sum += nums[i]==0?-1:1;
            
            if(sum==0)
            {   // because we start with index 0 so we will take i + 1 here 
                if(longest_subarray < i+1)
                    longest_subarray = i+1;
            }
            else if(mymap.find(sum)!=mymap.end())
            {
                if(longest_subarray < i-mymap[sum])
                    longest_subarray = i-mymap[sum];
            }
            else
                mymap[sum] = i;
        }
        
        return longest_subarray;
    }
};